Calculating and displaying status updates

julezjw opened the pull request:
Created functions to calculate status updates and commands that allow client to retrieve a status update.

julezjw committed:
Update Chat.java

julezjw committed:
Update UserContext.java

added USERstatusUpdate() which calls USERstatusUpdate() from User.java

julezjw committed:
Update User.java

added function to show userInterest updates

julezjw committed:
Create StatusUpdateTest.java

added a test file with one functional test

julezjw committed:
Update Chat.java

julezjw committed:
Update Controller.java

added statusUpdate()

julezjw committed:
Update NetworkCode.java

julezjw committed:
Update User.java

added commands for getting conversation interests

julezjw committed:
Update UserContext.java

julezjw commented on src/codeu/chat/common/User.java:
```diff
@@ -61,6 +67,16 @@ public User read(InputStream in) throws IOException {
+  public ArrayList<Uuid> userInterests = new ArrayList<Uuid>(); 
+  // convoInterests keeps track of user's CONVO interests with convo's uuid
+  // and the message count at the time of the last update 
+  public HashMap<Uuid, Integer> convoInterests = new HashMap<Uuid, Integer>();
```
this var is named conversationInterests in Priyanka's code

julezjw committed:
changed statusUpdate() to return updateTime

julezjw committed:
updates lastUpdateTime for every update

julezjw committed:
changes lastUpdateTime after update

julezjw committed:
changed var convoInterests->conversationInterests

julezjw committed:
initialized lastUpdateTime

TheMkrage commented on src/codeu/chat/client/commandline/Chat.java:
```diff
@@ -437,6 +440,43 @@ public void invoke(List<String> args) {
+			public void invoke(List<String> args) { 
+				HashMap<String, Integer> convoUpdates = user.CONVOstatusUpdate(); 
+				HashMap<Uuid, ArrayList<ArrayList<String>>> userUpdates = user.USERstatusUpdate();
+				int CONVOS_CREATED_ARRAY = 0; 
```
Maybe make these constants a public static field in UserContext, that way you do not need to declare them more than once.

TheMkrage replied:
Or in User

TheMkrage commented on src/codeu/chat/common/User.java:
```diff
@@ -104,4 +122,99 @@ private static int hash(User user) {
+  // newly created conversations (CONVOS_CREATED_ARRAY) and the second arraylist keeps 
+  // track of the conversations the userinterest has contributed messages to (CONVOS_CONTRIBUTED_TO_ARRAY)
+  //
+  public HashMap<Uuid, ArrayList<ArrayList<String>>> USERstatusUpdate(Iterable<ConversationContext> conversations, Time updateTime){ 
```
change to userStatusUpdate (unless there is a reason why its all caps USER that I am not aware of)

TheMkrage commented on src/codeu/chat/common/User.java:
```diff
@@ -104,4 +122,99 @@ private static int hash(User user) {
+			  convoUpdates.put(convo.conversation.title, numNewMessages); 
+			  
+			  // then update the value for the number of messages at lastUpdate in conversationInterests
+			  this.conversationInterests.put(convo.conversation.id, totalNumMessages); 
```
Does this remove the previous value? and is there anything in place to update the conversationIntersts Hashmap in the model?

julezjw replied:
It overwrites the previous value. And there aren't any functions in model atm; could you clarify why we would want to add those?

julezjw replied:
Ohh, i read your comments below and this made more sense. We would need to add these functions to model to keep things updated with persistent storage. Is that right?

TheMkrage commented on src/codeu/chat/client/core/UserContext.java:
```diff
@@ -53,4 +56,18 @@ public ConversationContext start(String name) {
+
+  // USERstatusUpdate()
+  public HashMap<Uuid, ArrayList<ArrayList<String>>> USERstatusUpdate(){ 
+	  HashMap<Uuid, ArrayList<ArrayList<String>>> userUpdates = this.user.USERstatusUpdate(this.conversations(), controller.statusUpdate()); 
```
Since there is no other opertations performed on userUpdate, maybe just `return this.user.USERstatusUpdate(this.conversations(), controller.statusUpdate());`

TheMkrage commented on src/codeu/chat/client/core/Controller.java:
```diff
@@ -112,4 +114,27 @@ public ConversationHeader newConversation(String title, Uuid owner)  {
+	try (final Connection connection = source.connect()) {
+
+	  Serializers.INTEGER.write(connection.out(), NetworkCode.STATUS_UPDATE_REQUEST);
+	  
```
might need to write an updated conversationInterests Hashmap to update the conversationInterests object of the model's users.

TheMkrage replied:
Actually on second thoughts, the code will work in a single session, but if we were to save the interests between sessions via json, we would need to find a way to update the convo interests in the model.

TheMkrage replied:
So for now its fine I think

TheMkrage reviewed (commented):
Functionally looks good to me . Except for the convo in model issue that we should address after this Monday. Just small stuff. Good job!

julezjw committed:
fixed some static ints 

made CONVOS_CREATED_ARRAY and CONVOS_CONTRIBUTED_TO_ARRAY static in the user class

julezjw committed:
got rid of some unnecessary code

julezjw committed:
adjusted for changes in static variables

Priya411 commented on src/codeu/chat/common/User.java:
```diff
@@ -104,4 +123,98 @@ private static int hash(User user) {
+  // newly created conversations (CONVOS_CREATED_ARRAY) and the second arraylist keeps 
+  // track of the conversations the userinterest has contributed messages to (CONVOS_CONTRIBUTED_TO_ARRAY)
+  //
+  public HashMap<Uuid, ArrayList<ArrayList<String>>> userStatusUpdate(Iterable<ConversationContext> conversations, Time updateTime){ 
```
Just wanted to check, is there a specific reason why updateTime is type Time and not type long?

KevinWorkman replied:
So the ArrayList of ArrayLists only contains two ArrayLists? I'd consider using a new class for this instead.

Priya411 commented on test/codeu/chat/server/StatusUpdateTest.java:
```diff
@@ -0,0 +1,156 @@
+
+
+    @Test
+    public void testUserInterestList() {
```
This seems like a good start on tests!

KevinWorkman commented on src/codeu/chat/client/core/Controller.java:
```diff
@@ -112,4 +114,27 @@ public ConversationHeader newConversation(String title, Uuid owner)  {
     return response;
   }
+  
+  public Time statusUpdate() {
```
I'd add comments to this function. When exactly is it called? What exactly is it returning?

KevinWorkman commented on src/codeu/chat/client/core/UserContext.java:
```diff
@@ -53,4 +56,16 @@ public ConversationContext start(String name) {
+  
+
+  // USERstatusUpdate()
+  public HashMap<Uuid, ArrayList<ArrayList<String>>> userStatusUpdate(){ 
```
I would add comments to these functions explaining exactly what the data structures contain.

KevinWorkman commented on src/codeu/chat/common/User.java:
```diff
@@ -104,4 +123,98 @@ private static int hash(User user) {
+	  // creating the dictionary userUpdates
+	  for (Uuid uuid : this.userInterests){
+		  userUpdates.put(uuid, new ArrayList<ArrayList<String>>(Arrays.asList(
+				  new ArrayList<String>() {}, new ArrayList<String>() {}))); 
```
I would think about converting this into a class instead of an ArrayList that contains two ArrayLists. something like this:

```
class UserInterest{
   ArrayList<String> convosCreated;
   ArrayList<String> convosContributedTo;
}
```

Then your map would be:

```
HashMap<Uuid, UserInterest> userUpdates;
```

KevinWorkman commented on src/codeu/chat/common/User.java:
```diff
@@ -104,4 +123,98 @@ private static int hash(User user) {
+	  }
+	  
+	  // looping through every conversation in the chat to organize updates 
+	  for (ConversationContext convo : conversations){
```
Just something to think about: what if you had a million conversations?

KevinWorkman commented on src/codeu/chat/common/User.java:
```diff
@@ -104,4 +123,98 @@ private static int hash(User user) {
+				  // If so, we add it to the user's CONVOS_CONTRIBUTED_TO_ARRAY, 
+				  // otherwise we continue iterating through the messages in the convo 
+				  if (this.userInterests.contains(currentmessage.message.author) && 
+						  !userUpdates.get(convo.conversation.owner.id()).get(CONVOS_CONTRIBUTED_TO_ARRAY).contains(convo.conversation.title)){
```
Another thing just to think about: what if the user is interested in a million conversations? The `contains()` function has to loop through all of them, and this is happening for every possible conversation. Not saying you have to change it now, but scalability is Google's favorite question, so it's good to start thinking about issues like that.

KevinWorkman reviewed (approved):
Looks good. Couple comments below.

It's up to you whether you want to try to address the comments or focus on integrating all of the changes, so I'm going to approve this now!

TheMkrage committed:
Merge branch 'master' into jw
